File: secret_diary.go
// secret_diary.go — Go версия

package main

import (
	"bufio"
	"crypto/sha256"
	"encoding/csv"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

type Entry struct {
	ID       int    `json:"id"`
	Date     string `json:"date"`
	Category string `json:"category"`
	Text     string `json:"text"`
}

type Diary struct {
	PinHash   string  `json:"pin_hash"`
	Entries   []Entry `json:"entries"`
	Locked    bool
	LastActivity time.Time
	file      string
}

func NewDiary(file string) *Diary {
	d := &Diary{file: file, Locked: true, LastActivity: time.Now()}
	d.load()
	return d
}

func (d *Diary) load() {
	data, err := os.ReadFile(d.file)
	if err != nil {
		d.PinHash = ""
		d.Entries = []Entry{}
		return
	}
	json.Unmarshal(data, d)
}

func (d *Diary) save() {
	data, _ := json.MarshalIndent(d, "", "  ")
	os.WriteFile(d.file, data, 0644)
}

func (d *Diary) hashPin(pin string) string {
	hash := sha256.Sum256([]byte(pin))
	return hex.EncodeToString(hash[:])
}

func (d *Diary) checkPin(pin string) bool {
	if d.PinHash == "" {
		d.PinHash = d.hashPin(pin)
		d.save()
		return true
	}
	return d.hashPin(pin) == d.PinHash
}

func (d *Diary) changePin(oldPin, newPin string) bool {
	if d.checkPin(oldPin) {
		d.PinHash = d.hashPin(newPin)
		d.save()
		return true
	}
	return false
}

func (d *Diary) unlock(pin string) bool {
	if d.checkPin(pin) {
		d.Locked = false
		d.LastActivity = time.Now()
		return true
	}
	return false
}

func (d *Diary) checkLock() bool {
	if time.Since(d.LastActivity) > 300*time.Second {
		d.Locked = true
		return true
	}
	return false
}

func (d *Diary) addEntry(category, text string) int {
	id := len(d.Entries) + 1
	date := time.Now().Format("2006-01-02 15:04:05")
	d.Entries = append(d.Entries, Entry{ID: id, Date: date, Category: category, Text: text})
	d.save()
	d.LastActivity = time.Now()
	return id
}

func (d *Diary) listAll() {
	if len(d.Entries) == 0 {
		fmt.Println("\u001B[33mНет записей.\u001B[0m")
		return
	}
	fmt.Printf("\u001B[36m%-4s %-20s %-12s %-50s\u001B[0m\n", "ID", "Дата", "Категория", "Текст")
	fmt.Println(strings.Repeat("-", 90))
	for _, e := range d.Entries {
		text := e.Text
		if len(text) > 50 {
			text = text[:50]
		}
		fmt.Printf("%-4d %-20s %-12s %-50s\n", e.ID, e.Date, e.Category, text)
	}
}

func (d *Diary) search(keyword string) {
	results := []Entry{}
	for _, e := range d.Entries {
		if strings.Contains(strings.ToLower(e.Text), strings.ToLower(keyword)) {
			results = append(results, e)
		}
	}
	if len(results) == 0 {
		fmt.Println("\u001B[33mНичего не найдено.\u001B[0m")
		return
	}
	for _, e := range results {
		fmt.Printf("%d: %s | %s | %s\n", e.ID, e.Date, e.Category, e.Text)
	}
}

func (d *Diary) delete(id int) bool {
	for i, e := range d.Entries {
		if e.ID == id {
			d.Entries = append(d.Entries[:i], d.Entries[i+1:]...)
			d.save()
			d.LastActivity = time.Now()
			return true
		}
	}
	return false
}

func (d *Diary) edit(id int, newText string) bool {
	for i, e := range d.Entries {
		if e.ID == id {
			d.Entries[i].Text = newText
			d.save()
			d.LastActivity = time.Now()
			return true
		}
	}
	return false
}

func (d *Diary) exportCSV(filename string) {
	if len(d.Entries) == 0 {
		fmt.Println("\u001B[33mНет записей для экспорта.\u001B[0m")
		return
	}
	file, _ := os.Create(filename)
	defer file.Close()
	writer := csv.NewWriter(file)
	defer writer.Flush()
	writer.Write([]string{"ID", "Дата", "Категория", "Текст"})
	for _, e := range d.Entries {
		writer.Write([]string{strconv.Itoa(e.ID), e.Date, e.Category, e.Text})
	}
	fmt.Printf("\u001B[32m💾 Экспорт CSV: %s\u001B[0m\n", filename)
}

func (d *Diary) exportJSON(filename string) {
	if len(d.Entries) == 0 {
		fmt.Println("\u001B[33mНет записей для экспорта.\u001B[0m")
		return
	}
	data, _ := json.MarshalIndent(d.Entries, "", "  ")
	os.WriteFile(filename, data, 0644)
	fmt.Printf("\u001B[32m💾 Экспорт JSON: %s\u001B[0m\n", filename)
}

func main() {
	diary := NewDiary("secret_diary.json")
	reader := bufio.NewReader(os.Stdin)

	fmt.Println("\u001B[36m🔐 Secret Diary (Go)\u001B[0m")

	if diary.PinHash == "" {
		fmt.Println("\u001B[33mПервый запуск! Установите PIN-код (4-6 цифр):\u001B[0m")
		for {
			fmt.Print("PIN: ")
			pin, _ := reader.ReadString('\n')
			pin = strings.TrimSpace(pin)
			if len(pin) >= 4 && len(pin) <= 6 && isDigit(pin) {
				fmt.Print("Повторите PIN: ")
				confirm, _ := reader.ReadString('\n')
				confirm = strings.TrimSpace(confirm)
				if pin == confirm {
					diary.PinHash = diary.hashPin(pin)
					diary.save()
					fmt.Println("\u001B[32m✅ PIN установлен!\u001B[0m")
					diary.Locked = false
					break
				} else {
					fmt.Println("\u001B[31m❌ PIN не совпадают.\u001B[0m")
				}
			} else {
				fmt.Println("\u001B[31m❌ PIN должен быть 4-6 цифр.\u001B[0m")
			}
		}
	}

	attempts := 3
	for diary.Locked {
		if attempts <= 0 {
			fmt.Println("\u001B[31m❌ Слишком много неудачных попыток. Выход.\u001B[0m")
			os.Exit(1)
		}
		fmt.Print("Введите PIN-код: ")
		pin, _ := reader.ReadString('\n')
		pin = strings.TrimSpace(pin)
		if diary.unlock(pin) {
			fmt.Println("\u001B[32m✅ Добро пожаловать!\u001B[0m")
			break
		} else {
			attempts--
			fmt.Printf("\u001B[31m❌ Неверный PIN. Осталось попыток: %d\u001B[0m\n", attempts)
		}
	}

	for {
		if diary.checkLock() {
			fmt.Println("\n\u001B[33m🔒 Автоматическая блокировка (бездействие > 5 мин)\u001B[0m")
			attempts = 3
			for diary.Locked {
				if attempts <= 0 {
					fmt.Println("\u001B[31m❌ Слишком много неудачных попыток. Выход.\u001B[0m")
					os.Exit(1)
				}
				fmt.Print("Введите PIN-код для разблокировки: ")
				pin, _ := reader.ReadString('\n')
				pin = strings.TrimSpace(pin)
				if diary.unlock(pin) {
					fmt.Println("\u001B[32m✅ Разблокировано!\u001B[0m")
					break
				} else {
					attempts--
					fmt.Printf("\u001B[31m❌ Неверный PIN. Осталось попыток: %d\u001B[0m\n", attempts)
				}
			}
		}

		fmt.Println("\n\u001B[36m🔐 Secret Diary (Go)\u001B[0m")
		fmt.Println("1. Добавить запись")
		fmt.Println("2. Показать все записи")
		fmt.Println("3. Поиск записей")
		fmt.Println("4. Редактировать запись")
		fmt.Println("5. Удалить запись")
		fmt.Println("6. Экспорт в CSV")
		fmt.Println("7. Экспорт в JSON")
		fmt.Println("8. Сменить PIN-код")
		fmt.Println("9. Выход")
		fmt.Print("Выберите действие: ")
		choice, _ := reader.ReadString('\n')
		choice = strings.TrimSpace(choice)

		switch choice {
		case "1":
			fmt.Print("Категория (личное/работа/идеи/другое): ")
			category, _ := reader.ReadString('\n')
			category = strings.TrimSpace(strings.ToLower(category))
			if category != "личное" && category != "работа" && category != "идеи" && category != "другое" {
				category = "другое"
			}
			fmt.Print("Текст записи: ")
			text, _ := reader.ReadString('\n')
			text = strings.TrimSpace(text)
			id := diary.addEntry(category, text)
			fmt.Printf("\u001B[32m✅ Запись добавлена (ID: %d)\u001B[0m\n", id)
		case "2":
			diary.listAll()
		case "3":
			fmt.Print("Ключевое слово для поиска: ")
			keyword, _ := reader.ReadString('\n')
			keyword = strings.TrimSpace(keyword)
			diary.search(keyword)
		case "4":
			diary.listAll()
			fmt.Print("Введите ID для редактирования: ")
			idStr, _ := reader.ReadString('\n')
			id, _ := strconv.Atoi(strings.TrimSpace(idStr))
			fmt.Print("Новый текст: ")
			text, _ := reader.ReadString('\n')
			text = strings.TrimSpace(text)
			if diary.edit(id, text) {
				fmt.Println("\u001B[32m✅ Запись обновлена.\u001B[0m")
			} else {
				fmt.Println("\u001B[31m❌ Запись не найдена.\u001B[0m")
			}
		case "5":
			diary.listAll()
			fmt.Print("Введите ID для удаления: ")
			idStr, _ := reader.ReadString('\n')
			id, _ := strconv.Atoi(strings.TrimSpace(idStr))
			if diary.delete(id) {
				fmt.Println("\u001B[32m✅ Запись удалена.\u001B[0m")
			} else {
				fmt.Println("\u001B[31m❌ Запись не найдена.\u001B[0m")
			}
		case "6":
			fmt.Print("Имя CSV файла (по умолч. diary_export.csv): ")
			filename, _ := reader.ReadString('\n')
			filename = strings.TrimSpace(filename)
			if filename == "" {
				filename = "diary_export.csv"
			}
			diary.exportCSV(filename)
		case "7":
			fmt.Print("Имя JSON файла (по умолч. diary_export.json): ")
			filename, _ := reader.ReadString('\n')
			filename = strings.TrimSpace(filename)
			if filename == "" {
				filename = "diary_export.json"
			}
			diary.exportJSON(filename)
		case "8":
			fmt.Print("Текущий PIN: ")
			old, _ := reader.ReadString('\n')
			old = strings.TrimSpace(old)
			fmt.Print("Новый PIN (4-6 цифр): ")
			newPin, _ := reader.ReadString('\n')
			newPin = strings.TrimSpace(newPin)
			if len(newPin) >= 4 && len(newPin) <= 6 && isDigit(newPin) {
				fmt.Print("Повторите новый PIN: ")
				confirm, _ := reader.ReadString('\n')
				confirm = strings.TrimSpace(confirm)
				if newPin == confirm {
					if diary.changePin(old, newPin) {
						fmt.Println("\u001B[32m✅ PIN изменён.\u001B[0m")
					} else {
						fmt.Println("\u001B[31m❌ Неверный текущий PIN.\u001B[0m")
					}
				} else {
					fmt.Println("\u001B[31m❌ PIN не совпадают.\u001B[0m")
				}
			} else {
				fmt.Println("\u001B[31m❌ PIN должен быть 4-6 цифр.\u001B[0m")
			}
		case "9":
			fmt.Println("До свидания!")
			return
		default:
			fmt.Println("\u001B[31mНеверный выбор.\u001B[0m")
		}
	}
}

func isDigit(s string) bool {
	for _, c := range s {
		if c < '0' || c > '9' {
			return false
		}
	}
	return true
}

File: secret_diary.cs
// secret_diary.cs — C# версия

using System;
using System.Collections.Generic;
using System.IO;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

class Entry {
    public int Id { get; set; }
    public string Date { get; set; }
    public string Category { get; set; }
    public string Text { get; set; }
}

class SecretDiary {
    private List<Entry> entries = new List<Entry>();
    private string pinHash = null;
    private bool locked = true;
    private DateTime lastActivity = DateTime.Now;
    private const string DataFile = "secret_diary.json";
    private const int LockTimeout = 300; // 5 минут

    public SecretDiary() {
        Load();
    }

    private void Load() {
        if (File.Exists(DataFile)) {
            try {
                string json = File.ReadAllText(DataFile);
                var data = JsonSerializer.Deserialize<Dictionary<string, object>>(json);
                if (data != null) {
                    pinHash = data.ContainsKey("pinHash") ? data["pinHash"].ToString() : null;
                    var entriesData = data.ContainsKey("entries") ? data["entries"].ToString() : "[]";
                    entries = JsonSerializer.Deserialize<List<Entry>>(entriesData) ?? new List<Entry>();
                }
            } catch {
                pinHash = null;
                entries = new List<Entry>();
            }
        }
    }

    private void Save() {
        var data = new Dictionary<string, object> {
            ["pinHash"] = pinHash,
            ["entries"] = entries
        };
        string json = JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(DataFile, json);
    }

    private string HashPin(string pin) {
        using var sha256 = SHA256.Create();
        byte[] hash = sha256.ComputeHash(Encoding.UTF8.GetBytes(pin));
        return BitConverter.ToString(hash).Replace("-", "").ToLower();
    }

    private bool CheckPin(string pin) {
        if (string.IsNullOrEmpty(pinHash)) {
            pinHash = HashPin(pin);
            Save();
            return true;
        }
        return HashPin(pin) == pinHash;
    }

    public bool Unlock(string pin) {
        if (CheckPin(pin)) {
            locked = false;
            lastActivity = DateTime.Now;
            return true;
        }
        return false;
    }

    public bool CheckLock() {
        if ((DateTime.Now - lastActivity).TotalSeconds > LockTimeout) {
            locked = true;
            return true;
        }
        return false;
    }

    public bool ChangePin(string oldPin, string newPin) {
        if (CheckPin(oldPin)) {
            pinHash = HashPin(newPin);
            Save();
            return true;
        }
        return false;
    }

    public int AddEntry(string category, string text) {
        int id = entries.Count + 1;
        string date = DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss");
        entries.Add(new Entry { Id = id, Date = date, Category = category, Text = text });
        Save();
        lastActivity = DateTime.Now;
        return id;
    }

    public void ListAll() {
        if (entries.Count == 0) {
            Console.WriteLine("\u001B[33mНет записей.\u001B[0m");
            return;
        }
        Console.WriteLine($"\u001B[36m{"ID",-4} {"Дата",-20} {"Категория",-12} {"Текст",-50}\u001B[0m");
        Console.WriteLine(new string('-', 90));
        foreach (var e in entries) {
            string text = e.Text.Length > 50 ? e.Text.Substring(0, 50) : e.Text;
            Console.WriteLine($"{e.Id,-4} {e.Date,-20} {e.Category,-12} {text,-50}");
        }
    }

    public void Search(string keyword) {
        var results = entries.FindAll(e => e.Text.ToLower().Contains(keyword.ToLower()));
        if (results.Count == 0) {
            Console.WriteLine("\u001B[33mНичего не найдено.\u001B[0m");
            return;
        }
        foreach (var e in results) {
            Console.WriteLine($"{e.Id}: {e.Date} | {e.Category} | {e.Text}");
        }
    }

    public bool Delete(int id) {
        var entry = entries.Find(e => e.Id == id);
        if (entry != null) {
            entries.Remove(entry);
            Save();
            lastActivity = DateTime.Now;
            return true;
        }
        return false;
    }

    public bool Edit(int id, string newText) {
        var entry = entries.Find(e => e.Id == id);
        if (entry != null) {
            entry.Text = newText;
            Save();
            lastActivity = DateTime.Now;
            return true;
        }
        return false;
    }

    public void ExportCSV(string filename = "diary_export.csv") {
        if (entries.Count == 0) {
            Console.WriteLine("\u001B[33mНет записей для экспорта.\u001B[0m");
            return;
        }
        using var writer = new StreamWriter(filename);
        writer.WriteLine("ID,Дата,Категория,Текст");
        foreach (var e in entries) {
            writer.WriteLine($"{e.Id},{e.Date},{e.Category},\"{e.Text}\"");
        }
        Console.WriteLine($"\u001B[32m💾 Экспорт CSV: {filename}\u001B[0m");
    }

    public void ExportJSON(string filename = "diary_export.json") {
        if (entries.Count == 0) {
            Console.WriteLine("\u001B[33mНет записей для экспорта.\u001B[0m");
            return;
        }
        string json = JsonSerializer.Serialize(entries, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(filename, json);
        Console.WriteLine($"\u001B[32m💾 Экспорт JSON: {filename}\u001B[0m");
    }

    private static bool IsDigit(string s) {
        foreach (char c in s) {
            if (!char.IsDigit(c)) return false;
        }
        return true;
    }

    public static void Main() {
        var diary = new SecretDiary();
        Console.WriteLine("\u001B[36m🔐 Secret Diary (C#)\u001B[0m");

        if (string.IsNullOrEmpty(diary.pinHash)) {
            Console.WriteLine("\u001B[33mПервый запуск! Установите PIN-код (4-6 цифр):\u001B[0m");
            while (true) {
                Console.Write("PIN: ");
                string pin = Console.ReadLine().Trim();
                if (IsDigit(pin) && pin.Length >= 4 && pin.Length <= 6) {
                    Console.Write("Повторите PIN: ");
                    string confirm = Console.ReadLine().Trim();
                    if (pin == confirm) {
                        diary.pinHash = diary.HashPin(pin);
                        diary.Save();
                        Console.WriteLine("\u001B[32m✅ PIN установлен!\u001B[0m");
                        diary.locked = false;
                        break;
                    } else {
                        Console.WriteLine("\u001B[31m❌ PIN не совпадают.\u001B[0m");
                    }
                } else {
                    Console.WriteLine("\u001B[31m❌ PIN должен быть 4-6 цифр.\u001B[0m");
                }
            }
        }

        int attempts = 3;
        while (diary.locked) {
            if (attempts <= 0) {
                Console.WriteLine("\u001B[31m❌ Слишком много неудачных попыток. Выход.\u001B[0m");
                Environment.Exit(1);
            }
            Console.Write("Введите PIN-код: ");
            string pin = Console.ReadLine().Trim();
            if (diary.Unlock(pin)) {
                Console.WriteLine("\u001B[32m✅ Добро пожаловать!\u001B[0m");
                break;
            } else {
                attempts--;
                Console.WriteLine($"\u001B[31m❌ Неверный PIN. Осталось попыток: {attempts}\u001B[0m");
            }
        }

        while (true) {
            if (diary.CheckLock()) {
                Console.WriteLine($"\n\u001B[33m🔒 Автоматическая блокировка (бездействие > {SecretDiary.LockTimeout} сек)\u001B[0m");
                attempts = 3;
                while (diary.locked) {
                    if (attempts <= 0) {
                        Console.WriteLine("\u001B[31m❌ Слишком много неудачных попыток. Выход.\u001B[0m");
                        Environment.Exit(1);
                    }
                    Console.Write("Введите PIN-код для разблокировки: ");
                    string pin = Console.ReadLine().Trim();
                    if (diary.Unlock(pin)) {
                        Console.WriteLine("\u001B[32m✅ Разблокировано!\u001B[0m");
                        break;
                    } else {
                        attempts--;
                        Console.WriteLine($"\u001B[31m❌ Неверный PIN. Осталось попыток: {attempts}\u001B[0m");
                    }
                }
            }

            Console.WriteLine("\n\u001B[36m🔐 Secret Diary (C#)\u001B[0m");
            Console.WriteLine("1. Добавить запись");
            Console.WriteLine("2. Показать все записи");
            Console.WriteLine("3. Поиск записей");
            Console.WriteLine("4. Редактировать запись");
            Console.WriteLine("5. Удалить запись");
            Console.WriteLine("6. Экспорт в CSV");
            Console.WriteLine("7. Экспорт в JSON");
            Console.WriteLine("8. Сменить PIN-код");
            Console.WriteLine("9. Выход");
            Console.Write("Выберите действие: ");
            string choice = Console.ReadLine().Trim();

            switch (choice) {
                case "1":
                    Console.Write("Категория (личное/работа/идеи/другое): ");
                    string category = Console.ReadLine().Trim().ToLower();
                    if (category != "личное" && category != "работа" && category != "идеи" && category != "другое") {
                        category = "другое";
                    }
                    Console.Write("Текст записи: ");
                    string text = Console.ReadLine().Trim();
                    int id = diary.AddEntry(category, text);
                    Console.WriteLine($"\u001B[32m✅ Запись добавлена (ID: {id})\u001B[0m");
                    break;
                case "2": diary.ListAll(); break;
                case "3":
                    Console.Write("Ключевое слово для поиска: ");
                    string keyword = Console.ReadLine().Trim();
                    diary.Search(keyword);
                    break;
                case "4":
                    diary.ListAll();
                    Console.Write("Введите ID для редактирования: ");
                    int editId = int.Parse(Console.ReadLine().Trim());
                    Console.Write("Новый текст: ");
                    string newText = Console.ReadLine().Trim();
                    if (diary.Edit(editId, newText)) {
                        Console.WriteLine("\u001B[32m✅ Запись обновлена.\u001B[0m");
                    } else {
                        Console.WriteLine("\u001B[31m❌ Запись не найдена.\u001B[0m");
                    }
                    break;
                case "5":
                    diary.ListAll();
                    Console.Write("Введите ID для удаления: ");
                    int delId = int.Parse(Console.ReadLine().Trim());
                    if (diary.Delete(delId)) {
                        Console.WriteLine("\u001B[32m✅ Запись удалена.\u001B[0m");
                    } else {
                        Console.WriteLine("\u001B[31m❌ Запись не найдена.\u001B[0m");
                    }
                    break;
                case "6": diary.ExportCSV(); break;
                case "7": diary.ExportJSON(); break;
                case "8":
                    Console.Write("Текущий PIN: ");
                    string old = Console.ReadLine().Trim();
                    Console.Write("Новый PIN (4-6 цифр): ");
                    string newPin = Console.ReadLine().Trim();
                    if (IsDigit(newPin) && newPin.Length >= 4 && newPin.Length <= 6) {
                        Console.Write("Повторите новый PIN: ");
                        string confirm = Console.ReadLine().Trim();
                        if (newPin == confirm) {
                            if (diary.ChangePin(old, newPin)) {
                                Console.WriteLine("\u001B[32m✅ PIN изменён.\u001B[0m");
                            } else {
                                Console.WriteLine("\u001B[31m❌ Неверный текущий PIN.\u001B[0m");
                            }
                        } else {
                            Console.WriteLine("\u001B[31m❌ PIN не совпадают.\u001B[0m");
                        }
                    } else {
                        Console.WriteLine("\u001B[31m❌ PIN должен быть 4-6 цифр.\u001B[0m");
                    }
                    break;
                case "9": Console.WriteLine("До свидания!"); return;
                default: Console.WriteLine("\u001B[31mНеверный выбор.\u001B[0m"); break;
            }
        }
    }
}
